Update FINEMAP to 1.4.2

aryarm opened the pull request:
Update [FINEMAP](https://bioconda.github.io/recipes/finemap/README.html): 1.4.1 → 1.4.2

----

Please read the [guidelines for Bioconda recipes](https://bioconda.github.io/contributor/guidelines.html) before opening a pull request (PR).

### General instructions

* If this PR adds or updates a recipe, use "Add" or "Update" appropriately as the first word in its title.
* New recipes not directly relevant to the biological sciences need to be submitted to the [conda-forge channel](https://conda-forge.org/docs/) instead of Bioconda.
* PRs require reviews prior to being merged. Once your PR is passing tests and ready to be merged, please issue the `@BiocondaBot please add label` command.
* Please post questions [on Gitter](https://gitter.im/bioconda/Lobby) or ping `@bioconda/core` in a comment.

### Instructions for avoiding API, ABI, and CLI breakage issues
Conda is able to record and lock (a.k.a. pin) dependency versions used at build time of other recipes.
This way, one can avoid that expectations of a downstream recipe with regards to API, ABI, or CLI are violated by later changes in the recipe.
If not already present in the meta.yaml, make sure to specify `run_exports` (see [here](https://bioconda.github.io/contributor/linting.html#missing-run-exports) for the rationale and comprehensive explanation).
Add a `run_exports` section like this:

```yaml
build:
  run_exports:
    - ...

```

with `...` being one of:

| Case                             | run_exports statement                                               |
| -------------------------------- | ------------------------------------------------------------------- |
| semantic versioning              | `{{ pin_subpackage("myrecipe", max_pin="x") }}`     |
| semantic versioning (0.x.x)      | `{{ pin_subpackage("myrecipe", max_pin="x.x") }}`   |
| known breakage in minor versions | `{{ pin_subpackage("myrecipe", max_pin="x.x") }}` (in such a case, please add a note that shortly mentions your evidence for that) |
| known breakage in patch versions | `{{ pin_subpackage("myrecipe", max_pin="x.x.x") }}` (in such a case, please add a note that shortly mentions your evidence for that) |
| calendar versioning              | `{{ pin_subpackage("myrecipe", max_pin=None) }}`    |

while replacing `"myrecipe"` with either `name` if a `name|lower` variable is defined in your recipe or with the lowercase name of the package in quotes.

### Bot commands for PR management

<details>
  <summary>Please use the following BiocondaBot commands:</summary>

Everyone has access to the following BiocondaBot commands, which can be given in a comment:

<table>
  <tr>
    <td><code>@BiocondaBot please update</code></td>
    <td>Merge the master branch into a PR.</td>
  </tr>
  <tr>
    <td><code>@BiocondaBot please add label</code></td>
    <td>Add the <code>please review & merge</code> label.</td>
  </tr>
  <tr>
    <td><code>@BiocondaBot please fetch artifacts</code></td>
    <td>Post links to CI-built packages/containers. <br />You can use this to test packages locally.</td>
  </tr>
</table>

Note that the <code>@BiocondaBot please merge</code> command is now depreciated. Please just squash and merge instead.

Also, the bot watches for comments from non-members that include `@bioconda/<team>` and will automatically re-post them to notify the addressed `<team>`.

</details>

aryarm commented:
@BiocondaBot please add label

I updated the recipe to work for linux, but I wasn't able to get it to work for osx too. The osx build is unable to find shared libraries. I tried using install_name_tool to replace the paths to the libraries, but I couldn't get it to work

martin-g commented on recipes/finemap/meta.yaml:
```diff
@@ -1,37 +1,44 @@
 source:
-  sha256: 75c919d9cc981bc08e1a982b1632110109ffe6a057e803c9aa36c143aa56d546
-  url: http://www.christianbenner.com/finemap_v1.4.1_x86_64.tgz
+  url: http://www.christianbenner.com/finemap_v{{ version }}_x86_64.tgz # [linux]
```
```suggestion
  - url: http://www.christianbenner.com/finemap_v{{ version }}_x86_64.tgz # [linux]
```
This looks wrong. The two pairs (url + sha256) should be two items of YAML list

martin-g commented on recipes/finemap/meta.yaml:
```diff
@@ -1,37 +1,44 @@
-  url: http://www.christianbenner.com/finemap_v1.4.1_x86_64.tgz
+  url: http://www.christianbenner.com/finemap_v{{ version }}_x86_64.tgz # [linux]
+  sha256: 3b1fc6eb3c2ccafd647b32e02d0244495cd0ade9ed7d474606c31ebf6e98b0c9 # [linux]
+  url: http://www.christianbenner.com/finemap_v{{ version }}_MacOSX.tgz # [osx]
```
```suggestion
  - url: http://www.christianbenner.com/finemap_v{{ version }}_MacOSX.tgz # [osx]
```

martin-g commented on recipes/finemap/meta.yaml:
```diff
@@ -1,37 +1,44 @@
 requirements:
-  build: []
+  build:
+    - {{ compiler('cxx') }}
```
Is the compiler needed if there is no build ? 
Above I see a script section that just copies the binaries to PREFIX/bin

aryarm replied:
I added the compiler here to get around the linter, which will otherwise complain that the package should be marked as noarch. Here's the error message I'm getting now
```
ERROR: recipes/finemap/meta.yaml:23: should_be_noarch_generic: The recipe should be build as `noarch`
```

Do you know if there's a better solution? I had borrowed this trick from the plink2 recipe:
https://github.com/bioconda/bioconda-recipes/blob/master/recipes/plink2/meta.yaml

martin-g replied:
IMO you should add a skip for the rules

aryarm replied:
ok, done! I removed the compiler requirement and added a few more dependencies that it seems like the compiler requirement must have been bringing in

I also had to add a few lint skips

aryarm commented on recipes/finemap/meta.yaml:
```diff
@@ -1,37 +1,38 @@
+  - url: http://www.christianbenner.com/finemap_v{{ version }}_MacOSX.tgz # [osx]
+    sha256: e2967e5586347a8ac5cbcddf3c0e41bf8d3e7793875f78b509d07903d17a415b # [osx]
+
 build:
-  binary_relocation: false
-  noarch: generic
-  number: '0'
+  number: 0
+  # skip broken osx for now
+  skip: True # [not linux64]
+  run_exports:
+    - {{ pin_subpackage('finemap', max_pin="x") }}
   script:
-    - install -Dd ${PREFIX}/bin
-    - install -D -m 755 finemap_v1.4.1_x86_64 ${PREFIX}/bin/finemap
-  string: '0'
-requirements:
-  build: []
-  host: []
-  run: []
+    - mkdir -p ${PREFIX}/bin
+    - install -m 755 finemap_v{{ version }}_x86_64 ${PREFIX}/bin/finemap # [linux]
+    - install -m 755 finemap_v{{ version }}_MacOSX ${PREFIX}/bin/finemap # [osx]
+
 test:
   commands:
-    - test -f $PREFIX/bin/finemap
+    - finemap --help
+
 about:
   home: http://www.christianbenner.com
   license: Other
   license_file: LICENSE
-  summary: Program for identifying causal SNPs and their effect sizes and heritability
-    contributions
+  summary: >
+    Program for identifying causal SNPs and their effect sizes and heritability contributions
+
```
Would it be sufficient to add this line in the about section?
```
license_url: http://www.christianbenner.com/license_finemap_v1.4.html
```

aryarm commented:
@BiocondaBot please fetch artifacts

aryarm commented:
@martin-g , this is ready for another review, if you're able 🙌 

I removed the compiler requirement and replaced it with requirements for each of the dynamic shared libraries that conda-bld reported in the binary. I also added more thorough tests using the example data in the source.

I decided to give up on the possibility of a MacOS build. As far as I can tell, it would require that the binary be rebuilt, but the source code isn't published.

Lastly, I wanted to point out a few warnings that I didn't resolve:
- `WARNING :: get_rpaths_raw()=[] and patchelf=[''] disagree for /opt/conda/conda-bld/finemap_1731534642894/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/bin/finemap ::` ([link](https://github.com/bioconda/bioconda-recipes/actions/runs/11826336886/job/32952193576#step:7:288))
- `WARNING (finemap): run-exports library package conda-forge/linux-64::libgomp==14.2.0=h77fa898_1 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to build/ignore_run_exports)` ([link](https://github.com/bioconda/bioconda-recipes/actions/runs/11826336886/job/32952193576#step:7:297))
- `WARNING (finemap): run-exports library package conda-forge/noarch::sysroot_linux-64==2.17=h4a8ded7_18 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to build/ignore_run_exports)` ([link](https://github.com/bioconda/bioconda-recipes/actions/runs/11826336886/job/32952193576#step:7:298))

I'm genuinely unsure of how to fix the first warning. And for the other two, I wasn't sure if it was save to add those packages to `ignore_run_exports`. Please let me know what you think!

bgruening commented on recipes/finemap/meta.yaml:
```diff
@@ -1,37 +1,57 @@
-  run: []
+  build:
+    - {{ stdlib("c") }}
+    - libgomp
```
please use the `compiler` directive here.

aryarm replied:
Hi @bgruening ,

Is this what you're thinking? 

```suggestion
    - {{ compiler('cxx') }}
    - libgomp
```

I actually had something like this in f9bee23, but we decided to remove it in https://github.com/bioconda/bioconda-recipes/pull/51978#discussion_r1835275055, since I'm simply copying the binary and not building it. Shall I add it back?

bgruening replied:
Yes this is what I was thinking. If @martin-g is ok with the way you have it now let's keep it. 

Adding libgomp and libgcc will be done by the complier directive and it's a bit more portable. But you eitherway skip everything but linux64. So 🤷

martin-g replied:
I'll take a closer look later today!

martin-g replied:
Do you need anything in the `requirements > build` section if you don't build it from source ?

aryarm replied:
The tests still seem to pass when I remove the `requirements > build` section, but the build will issue a bunch of warnings, which I was concerned about.
```
WARNING (finemap,bin/finemap): /lib64/libdl.so.2 not found in packages, sysroot(s) nor the missing_dso_whitelist... is this binary repackaging?
WARNING (finemap,bin/finemap): /lib64/libm.so.6 not found in packages, sysroot(s) nor the missing_dso_whitelist... is this binary repackaging?
WARNING (finemap,bin/finemap): $RPATH/libgomp.so.1 not found in packages, sysroot(s) nor the missing_dso_whitelist... is this binary repackaging?
WARNING (finemap,bin/finemap): /lib64/libgcc_s.so.1 not found in packages, sysroot(s) nor the missing_dso_whitelist... is this binary repackaging?
WARNING (finemap,bin/finemap): /lib64/libpthread.so.0 not found in packages, sysroot(s) nor the missing_dso_whitelist... is this binary repackaging?
WARNING (finemap,bin/finemap): /lib64/libc.so.6 not found in packages, sysroot(s) nor the missing_dso_whitelist... is this binary repackaging?
WARNING (finemap,bin/finemap): /lib64/ld-linux-x86-64.so.2 not found in packages, sysroot(s) nor the missing_dso_whitelist... is this binary repackaging?
```

martin-g replied:
I don't have any experience with this kind of problems but I think you need to move those deps to `requirements > host`

aryarm replied:
That makes sense. From what I've read online, shared libraries like those are supposed to be placed in the host section to ensure portability. 

But when I move those requirements into the host section, I still get the same warnings mentioned above. So I guess I'll add them to `missing_dso_whitelist`

bgruening reviewed (approved):
I wish we could just compile it from source. This looks very hacky to me. Maybe we can try compiling from source for the next release?

aryarm commented:
Thanks for merging, @bgruening ! Agreed that this seemed very hacky.

I can try to email the author next time and ask for the source code to be published. It would be much easier that way.

bgruening commented:
That would be nice. Thanks @aryarm
